pkg/redirectpolicy/manager.go

```diff
@@ -0,0 +1,531 @@
+			rpm.pendingEvents.forService[config.serviceID.String()] = &config
+		}
+	}
+	pods := rpm.getLocalPodsForPolicy(&config, rpm.epManager)
+	if len(pods) == 0 {
+		return true, nil
+	}
+	backends := rpm.getPodBackendsForPolicy(&config, pods...)
+	if len(backends) > 0 {
+		rpm.updatePolicyConfigBackends(&config, true, backends...)
+	}
+
+	return true, nil
+}
+
+// DeleteRedirectPolicy deletes the given local redirect policy config, and deletes
+// the internal state associated with the policy.
+func (rpm *RedirectPolicyManager) DeleteRedirectPolicy(config LRPConfig) (bool, error) {
+	rpm.Lock()
+	defer rpm.Unlock()
+
+	storedConfig := rpm.policyConfigs[config.getPolicyID()]
+	if storedConfig == nil {
+		return false, fmt.Errorf("Local redirect policy to be deleted not found")
+	}
+	if storedConfig.serviceID != nil {
+		rpm.deleteService(*storedConfig.serviceID)
+		delete(rpm.pendingEvents.forService, storedConfig.serviceID.String())
+	} else {
+		rpm.notifyPolicyFrontendDelete(storedConfig)
+	}
+	delete(rpm.policyConfigs, storedConfig.getPolicyID())
+	return false, nil
+}
+
+// OnAddService handles Kubernetes service updates, and updates the internal state
+// for the policy config that might be associated with the service.
+func (rpm *RedirectPolicyManager) OnAddService(svcID k8s.ServiceID, svcCache *k8s.ServiceCache) {
+	rpm.Lock()
+	defer rpm.Unlock()
+	if len(rpm.policyConfigs) == 0 {
+		return
+	}
+
+	// Check if there are any pending events for this service.
+	if config, ok := rpm.pendingEvents.forService[svcID.String()]; ok {
+		svcAddr := rpm.getSvcFrontend(*config, svcCache)
+		if svcAddr != nil {
+			delete(rpm.pendingEvents.forService, svcID.String())
+			rpm.updatePolicyConfigFrontend(config, svcAddr, true)
+		}
+	}
+}
+
+// deleteService deletes config associated with the specified service.
+// Note: Needs to be called with redirect policy manager lock.
+func (rpm *RedirectPolicyManager) deleteService(svcID k8s.ServiceID) {
+	if rp, ok := rpm.services[svcID.String()]; ok {
+		// Get the policy config that selects this service.
+		config := rpm.policyConfigs[rp]
+		rpm.notifyPolicyFrontendDelete(config)
+		// Add it to the pending list in case the service is re-added.
+		rpm.pendingEvents.forService[svcID.String()] = config
+		delete(rpm.services, svcID.String())
+	}
+}
+
+// OnDeleteService handles Kubernetes service deletes, and deletes the internal state
+// for the policy config that might be associated with the service.
+func (rpm *RedirectPolicyManager) OnDeleteService(svcID k8s.ServiceID) {
+	rpm.Lock()
+	defer rpm.Unlock()
+	if len(rpm.policyConfigs) == 0 {
+		return
+	}
+
+	if rp, ok := rpm.services[svcID.String()]; ok {
+		// Get the policy config that selects this service.
+		config := rpm.policyConfigs[rp]
+		rpm.notifyPolicyFrontendDelete(config)
+		// Add it to the pending list in case the service is re-added.
+		rpm.pendingEvents.forService[svcID.String()] = config
+		delete(rpm.services, svcID.String())
+	}
```

Comment:
This seems like it's duplicated with `deleteService()`. Can we just call that function here?